File: data-structures-algorithms/data-structures/assignments/src/week1_basic_data_structures/brackets_in_code/check_brackets.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Stack;

class Bracket {
    Bracket(char type, int position) {
        this.type = type;
        this.position = position;
    }

    boolean Match(char c) {
        if (this.type == '[' && c == ']')
            return true;
        if (this.type == '{' && c == '}')
            return true;
        if (this.type == '(' && c == ')')
            return true;
        return false;
    }

    char type;
    int position;
}

class check_brackets {
    public static void main(String[] args) throws IOException {
        InputStreamReader input_stream = new InputStreamReader(System.in);
        BufferedReader reader = new BufferedReader(input_stream);
        String text = reader.readLine();

        Stack<Bracket> opening_brackets_stack = new Stack<>();
        for (int position = 0; position < text.length(); ++position) {
            char next = text.charAt(position);

            if (next == '(' || next == '[' || next == '{') {
                opening_brackets_stack.push(new Bracket(next, position + 1));
            }

            if (next == ')' || next == ']' || next == '}') {
                if (opening_brackets_stack.size() == 0) {
                    System.out.println(position + 1);
                    return;
                }
                Bracket bracket = opening_brackets_stack.pop();
                if (!bracket.Match(next)) {
                    System.out.println(position + 1);
                    return;
                }
            }
        }

        if (opening_brackets_stack.size() == 0) {
            System.out.println("Success");
        } else {
            System.out.println(opening_brackets_stack.pop().position);
        }
    }
}

File: data-structures-algorithms/data-structures/assignments/src/week1_basic_data_structures/1_brackets_in_code/check_brackets.py
# python3

from collections import namedtuple

Bracket = namedtuple("Bracket", ["char", "position"])


def are_matching(left, right):
    return (left + right) in ["()", "[]", "{}"]


def find_mismatch(text):
    opening_brackets_stack = []
    for i, next in enumerate(text):
        if next in "([{":
            opening_brackets_stack.append((next, i))
        if next in ")]}":
            if len(opening_brackets_stack) == 0:
                return i + 1
            left = opening_brackets_stack.pop(-1)
            if not are_matching(left[0], next):
                return i + 1
    return "Success" if len(opening_brackets_stack) == 0 else opening_brackets_stack.pop(-1)[1] + 1


def main():
    text = input()
    print(find_mismatch(text))


if __name__ == "__main__":
    main()
